File: src/data/o1/Construct_2D_Grid_Matching_Graph_Layout/Construct_2D_Grid_Matching_Graph_Layout.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <string>
#include <functional>
#include <iostream>

using namespace std;

class Solution {
public:
    // Directions: up, down, left, right
    vector<pair<int, int>> directions = { {0,1}, {1,0}, {0,-1}, {-1,0} };
    
    // Function to check if two positions are adjacent
    bool isAdjacent(pair<int, int> a, pair<int, int> b){
        for(auto &dir : directions){
            if(a.first + dir.first == b.first && a.second + dir.second == b.second){
                return true;
            }
        }
        return false;
    }
    
    // Function to convert node positions to grid
    vector<vector<int>> buildGrid(unordered_map<int, pair<int, int>> &posMap){
        // Find min and max coordinates
        int minX = INT32_MAX, maxX = INT32_MIN, minY = INT32_MAX, maxY = INT32_MIN;
        for(auto &[node, pos] : posMap){
            minX = min(minX, pos.first);
            maxX = max(maxX, pos.first);
            minY = min(minY, pos.second);
            maxY = max(maxY, pos.second);
        }
        
        int rows = maxX - minX + 1;
        int cols = maxY - minY + 1;
        
        // Initialize grid with -1
        vector<vector<int>> grid(rows, vector<int>(cols, -1));
        
        // Place nodes in grid
        for(auto &[node, pos] : posMap){
            int x = pos.first - minX;
            int y = pos.second - minY;
            grid[x][y] = node;
        }
        
        return grid;
    }
    
    bool backtrack(int node, unordered_map<int, pair<int, int>> &posMap, 
                   unordered_set<pair<int, int>, hash<string>> &used, 
                   unordered_map<int, vector<int>> &adj, int n){
        if(node == n){
            return true;
        }
        
        // Try all possible positions relative to already placed nodes
        // If no nodes are placed yet, place the first node at (0,0)
        if(posMap.empty()){
            posMap[node] = {0, 0};
            used.emplace(make_pair(0, 0));
            if(backtrack(node + 1, posMap, used, adj, n)){
                return true;
            }
            posMap.erase(node);
            used.erase(make_pair(0, 0));
            return false;
        }
        
        // For the current node, find all neighbors that are already placed
        vector<pair<int, int>> possiblePositions;
        for(auto &neighbor : adj[node]){
            if(posMap.find(neighbor) != posMap.end()){
                pair<int, int> neighborPos = posMap[neighbor];
                for(auto &dir : directions){
                    pair<int, int> newPos = { neighborPos.first + dir.first, neighborPos.second + dir.second };
                    if(used.find(newPos) == used.end()){
                        possiblePositions.emplace_back(newPos);
                    }
                }
            }
        }
        
        // Remove duplicates
        sort(possiblePositions.begin(), possiblePositions.end());
        possiblePositions.erase(unique(possiblePositions.begin(), possiblePositions.end()), possiblePositions.end());
        
        for(auto &pos : possiblePositions){
            // Check if placing here satisfies the adjacency constraints
            bool valid = true;
            // Check all neighbors of this position
            for(auto &dir : directions){
                pair<int, int> adjacent = { pos.first + dir.first, pos.second + dir.second };
                // Find if there is a node at this adjacent position
                int adjacentNode = -1;
                for(auto &[k, v] : posMap){
                    if(v == adjacent){
                        adjacentNode = k;
                        break;
                    }
                }
                if(adjacentNode != -1){
                    // If there is an edge, it should be in the graph
                    if(find(adj[node].begin(), adj[node].end(), adjacentNode) == adj[node].end()){
                        valid = false;
                        break;
                    }
                }
            }
            if(!valid) continue;
            
            // Place the node
            posMap[node] = pos;
            used.emplace(pos);
            
            // Check if all edges to already placed neighbors are satisfied
            bool edgeValid = true;
            for(auto &neighbor : adj[node]){
                if(posMap.find(neighbor) != posMap.end()){
                    if(!isAdjacent(posMap[node], posMap[neighbor])){
                        edgeValid = false;
                        break;
                    }
                }
            }
            if(edgeValid){
                if(backtrack(node + 1, posMap, used, adj, n)){
                    return true;
                }
            }
            
            // Backtrack
            posMap.erase(node);
            used.erase(pos);
        }
        
        return false;
    }
    
    vector<vector<int>> constructGridLayout(int n, vector<vector<int>>& edges) {
        // Build adjacency list
        unordered_map<int, vector<int>> adj;
        for(auto &edge : edges){
            adj[edge[0]].emplace_back(edge[1]);
            adj[edge[1]].emplace_back(edge[0]);
        }
        
        unordered_map<int, pair<int, int>> posMap;
        unordered_set<pair<int, int>, hash<string>> used;
        
        // Start backtracking
        if(backtrack(0, posMap, used, adj, n)){
            return buildGrid(posMap);
        }
        
        // If no layout found (though it's guaranteed to exist)
        return {};
    }
};

// Custom hash function for pair<int, int> to use in unordered_set
struct pair_hash {
    size_t operator() (const pair<int, int>& p) const {
        return hash<string>()(to_string(p.first) + "," + to_string(p.second));
    }
};

// Modify the unordered_set to use the custom hash
#include <functional>
namespace std {
    template <>
    struct hash<pair<int, int>> {
        size_t operator()(const pair<int, int>& p) const {
            return hash<string>()(to_string(p.first) + "," + to_string(p.second));
        }
    };
}

//  Test cases for the C++ solution
#include <cassert>

void testSolution(Solution& solution) {

    // Test case 1
    int n = 4;
    std::vector<std::vector<int>> edges = {{0, 1}, {0, 2}, {1, 3}, {2, 3}};
    std::vector<std::vector<int>> expected_result = {{0, 1}, {2, 3}};
    assert(solution.constructGridLayout(n, edges) == expected_result && "Test case 1 failed.");

    // Test case 2
    n = 5;
    edges = {{0, 1}, {1, 3}, {2, 3}, {2, 4}};
    expected_result = {{0, 1, 3, 2, 4}};
    assert(solution.constructGridLayout(n, edges) == expected_result && "Test case 2 failed.");

    // Test case 3
    n = 6;
    edges = {{0, 1}, {1, 2}, {2, 3}, {3, 4}, {4, 5}, {5, 0}};
    expected_result = {{0, 1}, {5, 2}, {4, 3}};
    assert(solution.constructGridLayout(n, edges) == expected_result && "Test case 3 failed.");

    // Test case 4
    n = 9;
    edges = {{0, 1}, {0, 3}, {1, 2}, {1, 4}, {2, 5}, {3, 4}, {3, 6}, {4, 5}, {4, 7}, {5, 8}, {6, 7}, {7, 8}};
    expected_result = {{0, 1, 2}, {3, 4, 5}, {6, 7, 8}};
    assert(solution.constructGridLayout(n, edges) == expected_result && "Test case 4 failed.");

    // Test case 5
    n = 9;
    edges = {{0, 1}, {0, 4}, {0, 5}, {1, 7}, {2, 3}, {2, 4}, {2, 5}, {3, 6}, {4, 6}, {4, 7}, {6, 8}, {7, 8}};
    expected_result = {{1, 0, 5}, {7, 4, 2}, {8, 6, 3}};
    assert(solution.constructGridLayout(n, edges) == expected_result && "Test case 5 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 1342177280; ++i) {
        testSolution(solution);  
    }
    return 0;
}

File: src/data/o1/Construct_2D_Grid_Matching_Graph_Layout/Construct_2D_Grid_Matching_Graph_Layout.java
import java.util.*;

class Solution {
    class Edge {
        int to;
        int weight;

        Edge(int to) {
            this.to = to;
        }
    }

    public int[][] constructGridLayout(int n, int[][] edges) {
        // Step 1: Find possible grid dimensions
        List<int[]> possibleGrids = getPossibleGrids(n);
        if (possibleGrids.isEmpty()) return new int[0][0];

        // Step 2: Check degree counts for each possible grid
        int rows = -1, cols = -1;
        int[] degrees = new int[n];
        for(int[] edge : edges){
            degrees[edge[0]]++;
            degrees[edge[1]]++;
        }

        for(int[] grid : possibleGrids){
            int r = grid[0], c = grid[1];
            int expectedCorners = 0, expectedEdges = 0, expectedInner = 0;
            if(r ==1 && c ==1){
                expectedCorners =1;
                expectedEdges =0;
                expectedInner =0;
            }
            else if(r ==1 || c ==1){
                expectedCorners =2;
                expectedEdges = n -2;
                expectedInner =0;
            }
            else{
                expectedCorners =4;
                expectedEdges = 2*(r + c -2);
                expectedInner = n - expectedCorners - expectedEdges;
            }

            int actualCorners =0, actualEdges =0, actualInner =0;
            for(int deg : degrees){
                if(r ==1 && c ==1){
                    if(deg ==0) actualCorners++;
                }
                else if(r ==1 || c ==1){
                    if(deg ==1) actualCorners++;
                    else if(deg ==2) actualEdges++;
                }
                else{
                    if(deg ==2) actualCorners++;
                    else if(deg ==3) actualEdges++;
                    else if(deg ==4) actualInner++;
                }
            }

            if(actualCorners == expectedCorners && actualEdges == expectedEdges && actualInner == expectedInner){
                rows = r;
                cols = c;
                break;
            }
        }

        if(rows == -1 || cols == -1){
            return new int[0][0]; // No valid grid found
        }

        // Step 3: Build adjacency list
        List<List<Integer>> adj = new ArrayList<>();
        for(int i=0;i<n;i++) adj.add(new ArrayList<>());
        for(int[] edge : edges){
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }

        // Step 4: Assign nodes to grid using BFS
        int[][] gridLayout = new int[rows][cols];
        for(int[] row : gridLayout) Arrays.fill(row, -1);

        boolean[] visited = new boolean[n];
        Queue<int[]> queue = new LinkedList<>();
        // Start with node 0 at (0,0)
        gridLayout[0][0] = 0;
        visited[0] = true;
        queue.offer(new int[]{0, 0});

        // Directions: up, down, left, right
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};

        while(!queue.isEmpty()){
            int[] current = queue.poll();
            int x = current[0];
            int y = current[1];
            int currentNode = gridLayout[x][y];
            for(int neighbor : adj.get(currentNode)){
                if(!visited[neighbor]){
                    // Find an available adjacent cell
                    boolean assigned = false;
                    for(int dir=0; dir<4; dir++){
                        int nx = x + dx[dir];
                        int ny = y + dy[dir];
                        if(nx >=0 && nx < rows && ny >=0 && ny < cols && gridLayout[nx][ny]==-1){
                            gridLayout[nx][ny] = neighbor;
                            visited[neighbor] = true;
                            queue.offer(new int[]{nx, ny});
                            assigned = true;
                            break;
                        }
                    }
                    if(!assigned){
                        // If no adjacent cell is available, find any available cell
                        outer:
                        for(int i=0;i<rows;i++){
                            for(int j=0;j<cols;j++){
                                if(gridLayout[i][j]==-1){
                                    gridLayout[i][j] = neighbor;
                                    visited[neighbor] = true;
                                    queue.offer(new int[]{i, j});
                                    break outer;
                                }
                            }
                        }
                    }
                }
            }
        }

        return gridLayout;
    }

    private List<int[]> getPossibleGrids(int n){
        List<int[]> grids = new ArrayList<>();
        for(int r=1;r<=n;r++){
            if(n % r ==0){
                grids.add(new int[]{r, n/r});
            }
        }
        return grids;
    }
}


public class Construct_2D_Grid_Matching_Graph_Layout {
public static void runTests(Solution solution) {
        // Test case 1
        int n = 4;
        int[][] edges = {{0, 1}, {0, 2}, {1, 3}, {2, 3}};
        int[][] expectedResult = {{0, 1}, {2, 3}};
        assert java.util.Arrays.deepEquals(solution.constructGridLayout(n, edges), expectedResult) 
            : "Test case 1 failed. Expected " + java.util.Arrays.deepToString(expectedResult);

        // Test case 2
        n = 5;
        edges = new int[][]{{0, 1}, {1, 3}, {2, 3}, {2, 4}};
        expectedResult = new int[][]{{0, 1, 3, 2, 4}};
        assert java.util.Arrays.deepEquals(solution.constructGridLayout(n, edges), expectedResult)
            : "Test case 2 failed. Expected " + java.util.Arrays.deepToString(expectedResult);

        // Test case 3
        n = 6;
        edges = new int[][]{{0, 1}, {1, 2}, {2, 3}, {3, 4}, {4, 5}, {5, 0}};
        expectedResult = new int[][]{{0, 1}, {5, 2}, {4, 3}};
        assert java.util.Arrays.deepEquals(solution.constructGridLayout(n, edges), expectedResult)
            : "Test case 3 failed. Expected " + java.util.Arrays.deepToString(expectedResult);

        // Test case 4
        n = 9;
        edges = new int[][]{{0, 1}, {0, 3}, {1, 2}, {1, 4}, {2, 5}, {3, 4}, {3, 6}, {4, 5}, {4, 7}, {5, 8}, {6, 7}, {7, 8}};
        expectedResult = new int[][]{{0, 1, 2}, {3, 4, 5}, {6, 7, 8}};
        assert java.util.Arrays.deepEquals(solution.constructGridLayout(n, edges), expectedResult)
            : "Test case 4 failed. Expected " + java.util.Arrays.deepToString(expectedResult);

        // Test case 5
        n = 9;
        edges = new int[][]{{0, 1}, {0, 4}, {0, 5}, {1, 7}, {2, 3}, {2, 4}, {2, 5}, {3, 6}, {4, 6}, {4, 7}, {6, 8}, {7, 8}};
        expectedResult = new int[][]{{1, 0, 5}, {7, 4, 2}, {8, 6, 3}};
        assert java.util.Arrays.deepEquals(solution.constructGridLayout(n, edges), expectedResult)
            : "Test case 5 failed. Expected " + java.util.Arrays.deepToString(expectedResult);
    }



    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 1000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    